File: PokemonObjC/PokemonObjC/View Controllers/PokemonDetailVC/PokemonDetailViewController.m
//
//  PokemonDetailViewController.m
//  PokemonObjC
//
//  Created by Chris Gonzales on 4/24/20.
//  Copyright © 2020 Chris Gonzales. All rights reserved.
//

#import "PokemonDetailViewController.h"
#import "PokemonObjC-Swift.h"

@interface PokemonDetailViewController ()

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UILabel *idLabel;
@property (weak, nonatomic) IBOutlet UILabel *abilitiesLabel;

@end

@implementation PokemonDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [PokemonAPI.sharedController fillInDetailsFor:self.pokemon];
    [self.pokemon addObserver:self
                   forKeyPath:@"sprite"
                      options:0
                      context:nil];
    [self updateViews];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.pokemon removeObserver:self
                      forKeyPath:@"sprite"];
}

- (void)updateViews {
    if (self.pokemon) {
        
        self.title = [self.pokemon.name capitalizedString];
        
        if (self.pokemon.identifier) {
            self.idLabel.text = [NSString stringWithFormat:@"ID: %@",
                                 self.pokemon.identifier];
        }
        if (self.pokemon.abilities) {
            NSString *abilityString = [self.pokemon.abilities componentsJoinedByString:@", "];
            self.abilitiesLabel.text = [NSString stringWithFormat:@"Abilities: %@", abilityString];
        }
        if (self.pokemon.sprite) {
            NSURL *url = [NSURL URLWithString:self.pokemon.sprite];
            [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                if (data) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        self.imageView.image = [UIImage imageWithData:data];
                    });
                }
            }] resume];
        }
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:@"sprite"]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateViews];
        });
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

@end

File: Pokedex-Obj-C/Pokedex-Obj-C/View Controllers/PokemonDetailViewController.swift
//
//  PokemonDetailViewController.swift
//  Pokedex-Obj-C
//
//  Created by scott harris on 4/24/20.
//  Copyright © 2020 scott harris. All rights reserved.
//

import UIKit

class PokemonDetailViewController: UIViewController {
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var idLabel: UILabel!
    @IBOutlet weak var abilitiesLabel: UILabel!
    
    @objc var pokemon: SAHPokemon!
    var identifierObservation: NSKeyValueObservation?
    var imageDataObservation: NSKeyValueObservation?
    var abilitiesObservation: NSKeyValueObservation?
    
    override func viewDidLoad() {
        setupObservers()
        PokemonAPI.shared.fillInDetails(for: pokemon)
    }
    
    private func setupObservers() {
        identifierObservation = pokemon.observe(\.identifier, options: .new, changeHandler: { (object, change) in
            self.updateViews()
        })
        
        imageDataObservation = pokemon.observe(\.imageData, options: .new, changeHandler: { (object, change) in
            self.updateViews()
        })
        
        abilitiesObservation = pokemon.observe(\.abilities, options: .new, changeHandler: { (object, change) in
            self.updateViews()
        })

    }
    
    private func updateViews() {
        DispatchQueue.main.async {
            self.nameLabel.text = self.pokemon.name
            
            if let identifer = self.pokemon.identifier {
                self.idLabel.text = identifer.stringValue
            }
            
            if let imageData = self.pokemon.imageData {
                self.imageView.image = UIImage(data: imageData)
            }
            
            if let abilities = self.pokemon.abilities {
                var abilitesString = ""
                for ability in abilities {
                    abilitesString += ability
                    abilitesString += ", "
                }
                
                self.abilitiesLabel.text = abilitesString
            }
        }
        
    }
    
    deinit {
        identifierObservation = nil
        imageDataObservation = nil
        abilitiesObservation = nil
        
    }
    
}
